src/librustc_codegen_ssa/mir/rvalue.rs

```diff
@@ -327,13 +327,28 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
                                 if er.end != er.start
                                     && scalar.valid_range.end() > scalar.valid_range.start()
                                 {
-                                    // We want `table[e as usize]` to not
+                                    // We want `table[e as usize ± k]` to not
                                     // have bound checks, and this is the most
-                                    // convenient place to put the `assume`.
+                                    // convenient place to put the `assume`s.
                                     let ll_t_in_const =
                                         bx.cx().const_uint_big(ll_t_in, *scalar.valid_range.end());
-                                    let cmp = bx.icmp(IntPredicate::IntULE, llval, ll_t_in_const);
-                                    bx.assume(cmp);
+                                    let cmp_end =
+                                        bx.icmp(IntPredicate::IntULE, llval, ll_t_in_const);
+
+                                    if *scalar.valid_range.start() > 0 {
```

Comment:
Note to self/other reviewers: this condition should be enough because this is inside an `if` that checks for `scalar.valid_range.end() > scalar.valid_range.start()` (i.e. that the range *does not* wrap around, making it a valid simple *unsigned* integer range).

We could probably also support wraparound ranges, by treating them as *signed* integer ranges, but I guess that's less commonly used, especially for indexing?